File: analysis/eval_binop.py
# Ethan Doughty
# eval_binop.py
"""Binary operation shape inference."""

from __future__ import annotations
from typing import List, TYPE_CHECKING

from ir import Expr

import analysis.diagnostics as diag
from runtime.shapes import Shape, join_dim, dims_definitely_conflict
from analysis.constraints import record_constraint
from analysis.witness import ConflictSite

if TYPE_CHECKING:
    from analysis.diagnostics import Diagnostic


def eval_binop_ir(
    op: str,
    left: Shape,
    right: Shape,
    warnings: List['Diagnostic'],
    left_expr: Expr,
    right_expr: Expr,
    line: int,
    ctx,
    env
) -> Shape:
    """Evaluate a binary operation and infer result shape.

    Args:
        op: Operator string (+, -, *, .*, etc.)
        left: Left operand shape
        right: Right operand shape
        warnings: List to append warnings to
        left_expr: Left expression (for error messages)
        right_expr: Right expression (for error messages)
        line: Source line number
        ctx: Analysis context
        env: Current environment

    Returns:
        Result shape of the operation
    """
    if op in {"==", "~=", "<", "<=", ">", ">="}:
        if (left.is_matrix() and right.is_scalar()) or (left.is_scalar() and right.is_matrix()):
            warnings.append(diag.warn_suspicious_comparison_matrix_scalar(line, op, left_expr, right_expr, left, right)
            )
        elif left.is_matrix() and right.is_matrix():
            warnings.append(diag.warn_matrix_to_matrix_comparison(line, op, left_expr, right_expr, left, right)
            )
        return Shape.scalar()

    if op in {"&&", "||"}:
        if left.is_matrix() or right.is_matrix():
            warnings.append(diag.warn_logical_op_non_scalar(line, op, left_expr, right_expr, left, right)
            )
        return Shape.scalar()

    if op == ":":
        return Shape.matrix(1, None)

    # String arithmetic: string + string = numeric row vector (MATLAB behavior)
    if op == "+" and left.is_string() and right.is_string():
        return Shape.matrix(1, None)  # char + char = numeric row vector, length unknown

    # String + non-string: warning + unknown
    if op in {"+", "-", "*", ".*", "/", "./", ".^", "\\", "^"} and (left.is_string() or right.is_string()):
        if not (left.is_string() and right.is_string() and op == "+"):
            warnings.append(diag.warn_string_arithmetic(line, op, left, right))
            return Shape.unknown()

    # Type guard: non-numeric types cannot participate in arithmetic
    if not (left.is_unknown() or right.is_unknown()):
        if not left.is_numeric() or not right.is_numeric():
            warnings.append(diag.warn_arithmetic_type_mismatch(line, op, left_expr, right_expr, left, right))
            return Shape.unknown()

    if op == "^":
        # Matrix power: A^n (handle before generic scalar-broadcast shortcuts)
        # scalar^scalar -> scalar
        if left.is_scalar() and right.is_scalar():
            return Shape.scalar()
        # scalar^anything or anything^matrix -> unknown
        if left.is_scalar() or right.is_matrix():
            return Shape.unknown()
        # matrix^scalar: result same shape as A, but A must be square
        if left.is_matrix() and right.is_scalar():
            if dims_definitely_conflict(left.rows, left.cols):
                ctx.conflict_sites.append(ConflictSite(
                    dim_a=left.rows, dim_b=left.cols,
                    line=line, warning_code="W_MATRIX_POWER_NON_SQUARE",
                    constraints_snapshot=frozenset(ctx.constraints),
                    scalar_bindings_snapshot=tuple(sorted(ctx.scalar_bindings.items())),
                    value_ranges_snapshot=tuple(sorted(
                        (k, (v.lo, v.hi)) for k, v in ctx.value_ranges.items()
                    )),
                ))
                warnings.append(diag.warn_matrix_power_non_square(line, left_expr, left))
                return Shape.unknown()
            return left
        return Shape.unknown()

    if op == "\\":
        # mldivide: A\b — handle before generic scalar-broadcast shortcuts
        # scalar\scalar -> scalar
        if left.is_scalar() and right.is_scalar():
            return Shape.scalar()
        # scalar\matrix -> matrix (broadcast)
        if left.is_scalar() and not right.is_scalar():
            return right
        # matrix\scalar -> scalar
        if right.is_scalar() and left.is_matrix():
            return Shape.scalar()
        if left.is_unknown() or right.is_unknown():
            return Shape.unknown()
        if left.is_matrix() and right.is_matrix():
            # A[m x n] \ b[m x p] -> [n x p], require A.rows == b.rows
            record_constraint(ctx, env, left.rows, right.rows, line)
            if dims_definitely_conflict(left.rows, right.rows):
                ctx.conflict_sites.append(ConflictSite(
                    dim_a=left.rows, dim_b=right.rows,
                    line=line, warning_code="W_MLDIVIDE_DIM_MISMATCH",
                    constraints_snapshot=frozenset(ctx.constraints),
                    scalar_bindings_snapshot=tuple(sorted(ctx.scalar_bindings.items())),
                    value_ranges_snapshot=tuple(sorted(
                        (k, (v.lo, v.hi)) for k, v in ctx.value_ranges.items()
                    )),
                ))
                warnings.append(diag.warn_mldivide_dim_mismatch(line, left_expr, right_expr, left, right))
                return Shape.unknown()
            return Shape.matrix(left.cols, right.cols)
        return Shape.unknown()

    if left.is_scalar() and not right.is_scalar():
        return right
    if right.is_scalar() and not left.is_scalar():
        return left

    if op in {"+", "-", ".*", "./", "/", ".^"}:
        if left.is_unknown() or right.is_unknown():
            return Shape.unknown()

        if left.is_scalar() and right.is_scalar():
            # Division-by-zero check
            if op in {"/", "./"}:
                from analysis.eval_expr import _get_expr_interval
                from analysis.intervals import interval_is_exactly_zero

                divisor_iv = _get_expr_interval(right_expr, env, ctx)
                if interval_is_exactly_zero(divisor_iv):
                    warnings.append(diag.warn_division_by_zero(line, left_expr, right_expr))

            return Shape.scalar()

        if left.is_matrix() and right.is_matrix():
            # Record elementwise constraints before checking conflicts
            record_constraint(ctx, env, left.rows, right.rows, line)
            record_constraint(ctx, env, left.cols, right.cols, line)

            r_conflict = dims_definitely_conflict(left.rows, right.rows)
            c_conflict = dims_definitely_conflict(left.cols, right.cols)
            if r_conflict or c_conflict:
                # Record the first conflicting dim pair
                _da = left.rows if r_conflict else left.cols
                _db = right.rows if r_conflict else right.cols
                ctx.conflict_sites.append(ConflictSite(
                    dim_a=_da, dim_b=_db,
                    line=line, warning_code="W_ELEMENTWISE_MISMATCH",
                    constraints_snapshot=frozenset(ctx.constraints),
                    scalar_bindings_snapshot=tuple(sorted(ctx.scalar_bindings.items())),
                    value_ranges_snapshot=tuple(sorted(
                        (k, (v.lo, v.hi)) for k, v in ctx.value_ranges.items()
                    )),
                ))
                warnings.append(diag.warn_elementwise_mismatch(line, op, left_expr, right_expr, left, right)
                )
                return Shape.unknown()

            rows = join_dim(left.rows, right.rows)
            cols = join_dim(left.cols, right.cols)
            return Shape.matrix(rows, cols)

        return Shape.unknown()

    if op == "*":
        if left.is_scalar() and right.is_scalar():
            return Shape.scalar()

        if left.is_scalar() and right.is_matrix():
            return right
        if right.is_scalar() and left.is_matrix():
            return left

        if left.is_matrix() and right.is_matrix():
            # Record matrix multiplication inner dimension constraint
            record_constraint(ctx, env, left.cols, right.rows, line)

            if dims_definitely_conflict(left.cols, right.rows):
                ctx.conflict_sites.append(ConflictSite(
                    dim_a=left.cols, dim_b=right.rows,
                    line=line, warning_code="W_INNER_DIM_MISMATCH",
                    constraints_snapshot=frozenset(ctx.constraints),
                    scalar_bindings_snapshot=tuple(sorted(ctx.scalar_bindings.items())),
                    value_ranges_snapshot=tuple(sorted(
                        (k, (v.lo, v.hi)) for k, v in ctx.value_ranges.items()
                    )),
                ))
                suggest = (
                    not dims_definitely_conflict(left.rows, right.rows)
                    and not dims_definitely_conflict(left.cols, right.cols)
                )
                warnings.append(diag.warn_matmul_mismatch(line, left_expr, right_expr, left, right, suggest)
                )
                return Shape.unknown()

            return Shape.matrix(left.rows, right.cols)

        return Shape.unknown()

    if op in {"&", "|"}:
        # Element-wise logical: same shape rules as .* (broadcast compatible)
        if left.is_scalar() and right.is_scalar():
            return Shape.scalar()
        if left.is_scalar() and not right.is_scalar():
            return right
        if right.is_scalar() and not left.is_scalar():
            return left
        if left.is_unknown() or right.is_unknown():
            return Shape.unknown()
        if left.is_matrix() and right.is_matrix():
            record_constraint(ctx, env, left.rows, right.rows, line)
            record_constraint(ctx, env, left.cols, right.cols, line)
            r_conflict = dims_definitely_conflict(left.rows, right.rows)
            c_conflict = dims_definitely_conflict(left.cols, right.cols)
            if r_conflict or c_conflict:
                _da = left.rows if r_conflict else left.cols
                _db = right.rows if r_conflict else right.cols
                ctx.conflict_sites.append(ConflictSite(
                    dim_a=_da, dim_b=_db,
                    line=line, warning_code="W_ELEMENTWISE_MISMATCH",
                    constraints_snapshot=frozenset(ctx.constraints),
                    scalar_bindings_snapshot=tuple(sorted(ctx.scalar_bindings.items())),
                    value_ranges_snapshot=tuple(sorted(
                        (k, (v.lo, v.hi)) for k, v in ctx.value_ranges.items()
                    )),
                ))
                warnings.append(diag.warn_elementwise_mismatch(line, op, left_expr, right_expr, left, right))
                return Shape.unknown()
            rows = join_dim(left.rows, right.rows)
            cols = join_dim(left.cols, right.cols)
            return Shape.matrix(rows, cols)
        return Shape.unknown()

    return Shape.unknown()

File: src/EvalBinop.fs
module EvalBinop

open Ir
open Shapes
open Env
open Context
open Diagnostics
open SharedTypes

// ---------------------------------------------------------------------------
// Binary operation shape inference.
// Port of analysis/eval_binop.py
// ---------------------------------------------------------------------------

/// evalBinopIr: evaluate a binary operation and infer result shape.
/// getDivisorInterval: callback to compute interval of divisor expression (for div-by-zero check).
let evalBinopIr
    (op: string)
    (left: Shape)
    (right: Shape)
    (warnings: ResizeArray<Diagnostic>)
    (leftExpr: Expr)
    (rightExpr: Expr)
    (line: int)
    (ctx: AnalysisContext)
    (env: Env)
    (getDivisorInterval: Expr -> Interval option)
    : Shape =

    // Comparison operators: return broadcast shape (logical array matching operand shape)
    if Set.contains op (Set.ofList ["=="; "~="; "<"; "<="; ">"; ">="]) then
        match left, right with
        | Matrix _, Scalar | Scalar, Matrix _ ->
            warnings.Add(warnSuspiciousComparisonMatrixScalar line op leftExpr rightExpr left right)
        | Matrix _, Matrix _ ->
            warnings.Add(warnMatrixToMatrixComparison line op leftExpr rightExpr left right)
        | _ -> ()
        // Return shape reflects MATLAB broadcast semantics: A > 0 where A is matrix[m x n]
        // returns a logical matrix[m x n], not a scalar.
        match left, right with
        | Scalar, Scalar -> Scalar
        | Matrix(r, c), Scalar | Scalar, Matrix(r, c) -> Matrix(r, c)
        | Matrix(r1, c1), Matrix(r2, c2) -> Matrix(joinDim r1 r2, joinDim c1 c2)
        | UnknownShape, _ | _, UnknownShape -> UnknownShape
        | _ -> Scalar  // string, struct, cell, etc: scalar boolean

    // Short-circuit logical operators
    elif op = "&&" || op = "||" then
        match left, right with
        | Matrix _, _ | _, Matrix _ ->
            warnings.Add(warnLogicalOpNonScalar line op leftExpr rightExpr left right)
        | _ -> ()
        Scalar

    // Colon range: always 1 x unknown
    elif op = ":" then
        Matrix(Concrete 1, Unknown)

    // String + string = numeric row vector (MATLAB behavior)
    elif op = "+" && isString left && isString right then
        Matrix(Concrete 1, Unknown)

    // String + non-string: warning + unknown
    elif Set.contains op (Set.ofList ["+"; "-"; "*"; ".*"; "/"; "./"; ".^"; "\\"; "^"]) &&
         (isString left || isString right) then
        if not (isString left && isString right && op = "+") then
            warnings.Add(warnStringArithmetic line op left right)
            UnknownShape
        else
            Matrix(Concrete 1, Unknown)   // fallthrough for string+string

    // Type guard: non-numeric types cannot participate in arithmetic
    elif not (isUnknown left || isUnknown right) &&
         Set.contains op (Set.ofList ["+"; "-"; "*"; ".*"; "/"; "./"; ".^"; "\\"; "^"; "&"; "|"]) &&
         (not (isNumeric left) || not (isNumeric right)) then
        warnings.Add(warnArithmeticTypeMismatch line op leftExpr rightExpr left right)
        UnknownShape

    // Matrix power: A^n
    elif op = "^" then
        match left, right with
        | Scalar, Scalar -> Scalar
        | Scalar, _ | _, Matrix _ -> UnknownShape
        | Matrix(r, c), Scalar ->
            if dimsDefinitelyConflict r c then
                warnings.Add(warnMatrixPowerNonSquare line leftExpr left)
                UnknownShape
            else left
        | _ -> UnknownShape

    // mldivide: A\b
    elif op = "\\" then
        match left, right with
        | Scalar, Scalar -> Scalar
        | Scalar, _ -> right
        | _, Scalar when isMatrix left -> Scalar
        | UnknownShape, _ | _, UnknownShape -> UnknownShape
        | Matrix(lr, lc), Matrix(rr, rc) ->
            Constraints.recordConstraint ctx env lr rr line
            let lr' = Constraints.resolveDim ctx lr
            let rr' = Constraints.resolveDim ctx rr
            if dimsDefinitelyConflict lr' rr' then
                warnings.Add(warnMldivideDimMismatch line leftExpr rightExpr left right)
                UnknownShape
            else Matrix(lc, rc)
        | _ -> UnknownShape

    // Scalar broadcasting
    elif isScalar left && not (isScalar right) then right
    elif isScalar right && not (isScalar left) then left

    // Elementwise operations: +, -, .*, ./, /, .^
    elif Set.contains op (Set.ofList ["+"; "-"; ".*"; "./"; "/"; ".^"]) then
        match left, right with
        | UnknownShape, _ | _, UnknownShape -> UnknownShape
        | Scalar, Scalar ->
            // Division-by-zero check
            if op = "/" || op = "./" then
                let divisorIv = getDivisorInterval rightExpr
                if Intervals.intervalIsExactlyZero divisorIv then
                    warnings.Add(warnDivisionByZero line leftExpr rightExpr)
            Scalar
        | Matrix(r1, c1), Matrix(r2, c2) ->
            Constraints.recordConstraint ctx env r1 r2 line
            Constraints.recordConstraint ctx env c1 c2 line
            let r1' = Constraints.resolveDim ctx r1
            let r2' = Constraints.resolveDim ctx r2
            let c1' = Constraints.resolveDim ctx c1
            let c2' = Constraints.resolveDim ctx c2
            let rConflict = dimsDefinitelyConflict r1' r2'
            let cConflict = dimsDefinitelyConflict c1' c2'
            if rConflict || cConflict then
                warnings.Add(warnElementwiseMismatch line op leftExpr rightExpr left right)
                UnknownShape
            else Matrix(joinDim r1' r2', joinDim c1' c2')
        | _ -> UnknownShape

    // Matrix multiplication: *
    elif op = "*" then
        match left, right with
        | Scalar, Scalar -> Scalar
        | Scalar, Matrix _ -> right
        | Matrix _, Scalar -> left
        | Matrix(r1, c1), Matrix(r2, c2) ->
            Constraints.recordConstraint ctx env c1 r2 line
            let c1' = Constraints.resolveDim ctx c1
            let r2' = Constraints.resolveDim ctx r2
            if dimsDefinitelyConflict c1' r2' then
                let suggest =
                    not (dimsDefinitelyConflict r1 r2) &&
                    not (dimsDefinitelyConflict c1 c2)
                warnings.Add(warnMatmulMismatch line leftExpr rightExpr left right suggest)
                UnknownShape
            else Matrix(r1, c2)
        | _ -> UnknownShape

    // Element-wise logical: &, |
    elif op = "&" || op = "|" then
        match left, right with
        | Scalar, Scalar -> Scalar
        | Scalar, _ -> right
        | _, Scalar -> left
        | UnknownShape, _ | _, UnknownShape -> UnknownShape
        | Matrix(r1, c1), Matrix(r2, c2) ->
            Constraints.recordConstraint ctx env r1 r2 line
            Constraints.recordConstraint ctx env c1 c2 line
            let r1' = Constraints.resolveDim ctx r1
            let r2' = Constraints.resolveDim ctx r2
            let c1' = Constraints.resolveDim ctx c1
            let c2' = Constraints.resolveDim ctx c2
            let rConflict = dimsDefinitelyConflict r1' r2'
            let cConflict = dimsDefinitelyConflict c1' c2'
            if rConflict || cConflict then
                warnings.Add(warnElementwiseMismatch line op leftExpr rightExpr left right)
                UnknownShape
            else Matrix(joinDim r1' r2', joinDim c1' c2')
        | _ -> UnknownShape

    else UnknownShape
